### PR TITLE
8321039: [lworld] Substitutability code must recognize new annotations

### PR DESCRIPTION
**Pull request by mlchung**
An initial implementation for substitutability test and partial method handle and var handle support for null-restricted types. JDK-8317767 will provide the full library support for null-restricted types.    

This patch handles null-restricted field reference declared in a non-flat class.  The default value of that field is lazily set by VM on read access.  `Unsafe::getReference` may return null for null-restricted non-flat field.  

A new `jdk.internal.value.ValueClass` is created to provide JVM entry points for value classes until public APIs are defined.     `jdk.internal.misc.newNullRestrictedArray` method is moved as well.

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Issue
 * [JDK-8321039](https://bugs.openjdk.org/browse/JDK-8321039): [lworld] Substitutability code must recognize new annotations (**Enhancement** - P3)

### Reviewers
 * [Frederic Parain](https://openjdk.org/census#fparain) (@fparain - Committer) ⚠️ Review applies to [e216c78e](https://git.openjdk.org/valhalla/pull/960/files/e216c78eba5e8c3c2236d936c0dbe9d15e156570)
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - Committer) ⚠️ Review applies to [e216c78e](https://git.openjdk.org/valhalla/pull/960/files/e216c78eba5e8c3c2236d936c0dbe9d15e156570)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/valhalla.git pull/960/head:pull/960` \
`$ git checkout pull/960`

Update a local copy of the PR: \
`$ git checkout pull/960` \
`$ git pull https://git.openjdk.org/valhalla.git pull/960/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 960`

View PR using the GUI difftool: \
`$ git pr show -t 960`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/valhalla/pull/960.diff">https://git.openjdk.org/valhalla/pull/960.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/valhalla/pull/960#issuecomment-1841662196)

### CODE REVIEW

File: src/java.base/share/classes/java/lang/invoke/DirectMethodHandle.java [hunk 1]
```diff
@@ -671,32 +670,35 @@ private static LambdaForm preparedFieldLambdaForm(MemberName m) {
...
                 : Arrays.asList(m, m.getInvocationType().basicType(), lform, lform.methodType());
         return lform;
     }
 
-    private static LambdaForm preparedFieldLambdaForm(byte formOp, boolean isVolatile, boolean isValue, boolean isFlatValue, Class<?> ftype) {
+    private static LambdaForm preparedFieldLambdaForm(byte formOp, boolean isVolatile,
+                                                      boolean isValue, boolean isFlat, boolean isNullRestricted, Class<?> ftype) {
         int ftypeKind = ftypeKind(ftype, isValue);
-        int afIndex = afIndex(formOp, isVolatile, isFlatValue, ftypeKind);
+        int afIndex = afIndex(formOp, isVolatile, isFlat, ftypeKind);
```
Comment (liach) [inline src/java.base/share/classes/java/lang/invoke/DirectMethodHandle.java:689]: The null-restriction info should be encoded in the cache key too.

File: src/java.base/share/classes/java/lang/invoke/X-VarHandle.java.template [hunk 2]
```diff
@@ -43,33 +46,39 @@ final class VarHandle$Type$s {
 #if[Object]
         final Class<?> fieldType;
 #end[Object]
+#if[Reference]
+        final boolean nullRestricted;
+#end[Reference]
 
-        FieldInstanceReadOnly(Class<?> receiverType, long fieldOffset{#if[Object]?, Class<?> fieldType}) {
-            this(receiverType, fieldOffset{#if[Object]?, fieldType}, FieldInstanceReadOnly.FORM, false);
+        FieldInstanceReadOnly(Class<?> receiverType, long fieldOffset{#if[Object]?, Class<?> fieldType, boolean nullRestricted}) {
```
Comment (liach) [inline src/java.base/share/classes/java/lang/invoke/X-VarHandle.java.template:53]: This `nullRestricted` can be conditional to references, too

File: test/jdk/valhalla/valuetypes/NullRestrictedTest.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    @ImplicitlyConstructible
+    static value class EmptyContainer {
+        @NullRestricted
+        MyValueEmpty empty = new MyValueEmpty();
+    }
+
+    @Test
+    public void lazyInitializedDefaultValue() {
+        // VM lazily sets the null-restricted non-flat field to zero default
+        assertTrue(new EmptyContainer() == EmptyContainer.default);
```
Comment (fparain) [inline test/jdk/valhalla/valuetypes/NullRestrictedTest.java:55]: The `.default` syntax is translated to `aconst_init` which is going to be removed.
May be use `ValueClass.zeroInstance(EmptyContainer.class)` instead.
Applies to all `.default` usage in this file.

File: make/data/hotspot-symbols/symbols-unix [hunk 4]
```diff
@@ -137,6 +137,7 @@ JVM_GetSystemPackage
 JVM_GetSystemPackages
 JVM_GetTemporaryDirectory
 JVM_GetVmArguments
+JVM_GetZeroInstance
```
[the remark this answers is not in the document]
Comment (mlchung) [reply make/data/hotspot-symbols/symbols-unix:140]: That's the term described in the JEP.    I prefer to come back and do the renaming later once the APIs are defined.

File: src/hotspot/share/classfile/javaClasses.hpp [hunk 5]
```diff
@@ -1301,8 +1301,9 @@ class java_lang_invoke_MemberName: AllStatic {
     MN_TRUSTED_FINAL         = 0x00200000, // trusted final field
     MN_HIDDEN_MEMBER         = 0x00400000, // @Hidden annotation detected
     MN_FLAT_FIELD            = 0x00800000, // flat field
-    MN_REFERENCE_KIND_SHIFT  = 24, // refKind
-    MN_REFERENCE_KIND_MASK   = 0x0F000000 >> MN_REFERENCE_KIND_SHIFT,
+    MN_NULL_RESTRICTED_FIELD = 0x01000000, // null-restricted fiel
```
[the remark this answers is not in the document]
Comment (mlchung) [reply src/hotspot/share/classfile/javaClasses.hpp]: Fixed

File: src/java.base/share/classes/java/lang/invoke/DirectMethodHandle.java (hunk 1 above)
Comment (mlchung) [reply src/java.base/share/classes/java/lang/invoke/DirectMethodHandle.java]: Fixed.

**Review by fparain — APPROVED** [review]
Thank you for fixing this so quickly.

**Review by RogerRiggs — APPROVED** [review]
LGTM

**Comment by mlchung** [comment]
I restored `Unsafe::uninitializedDefaultValue` which is used by serialization in the current prototype.  Will revisit this in future work.

**Pull request closed without merging**